queue/amqp.go

```diff
@@ -316,24 +317,35 @@ func (q *AMQPQueue) RepublishBuried() error {
 			// if there is nothing after all the retries (meaning: BuriedQueue is surely
 			// empty or any arriving jobs will have to wait to the next call).
 			if retries > buriedNonBlockingRetries {
-				return nil
+				break
 			}
 
 			time.Sleep(50 * time.Millisecond)
 			retries++
 			continue
 		}
 
-		if err = j.Ack(); err != nil {
-			return err
-		}
-
 		retries = 0
+		if comply(j) {
+			if err = j.Ack(); err != nil {
```

Comment:
You should Ack the job after sending it to the other queue, and try to NACK if the publish fails.

Reply:
I would also reject the jobs just after publishing error or failing compliance. Adding them to a slice to process afterwards can make some of them unprocessed if some error happens.

Reply:
btw. did you think about `dead-letter-exchange`. we can use it for _"non immediately retry"_.

Reply:
@jfontan I can't reject immediately because I'd send back the job to the queue it's being iterated, so we will end up stuck in an infinite loop.

@kuba, this queue is already the `dead-letter-exchange` queue, so we would have to configure a buried queue for the buried queue, and republish from the child buried queue to the parent buried queue.

Finally we will have the same situation between those two `buried queues`. Not sure about it :confused:

Reply:
So maybe we need an another `exchange` with counter (and maybe size limit). If we exceed some q's size or timeout messages will be enqueued again but `retry counter` will be increased. If we got e.g. 5 retries and it's still failing we can drop a message. Maybe it's hard to explain as a comment but maybe just as a suggestion this is what I mean: https://felipeelias.github.io/rabbitmq/2016/02/22/rabbitmq-exponential-backoff.html
This is what me and my colleague (Felipe) implemented long time ago to avoid dropping messages and handle backoff.